### PR TITLE
refactor: added API, CSV, and UI test [SCHOOL-115]

### PR DESCRIPTION
**Pull request by Camus1859**

### CODE REVIEW

File: cypress/integration/precise-rounding.js
Comment (themightychris) [inline cypress/integration/precise-rounding.js]: **suggestion (non-blocking):** It might be good to stash all these in a fixture file, see [these examples](https://github.com/cypress-io/cypress-example-recipes/tree/master/examples/fundamentals__fixtures) and [these docs](https://docs.cypress.io/api/commands/fixture)

File: cypress/integration/precise-rounding.js
Comment (themightychris) [inline cypress/integration/precise-rounding.js]: **suggestion (non-blocking):** We generally try not to get too engineer-y with tests, but this might be a good case for generating all these `it()` blocks within a loop through all the test cases loaded from a fixture file as I mentioned above. You might just put all the statements as a property inside there, like:

```json
[
  {
    "caseTitle": "Student 5 ELA Competency 7",
    "student": "student4",
    "contentArea": "ELA",
    "competency": "7",
    "baseline": "6",
    "growth": "0.4",
    "progress": "63",
    "performanceLevel": "6.4"
  },
  {
    // ...
  }
]
```

File: cypress/integration/precise-rounding.js
Comment (themightychris) [inline cypress/integration/precise-rounding.js]: **suggestion (non-blocking)**: it should be possible to just declare these as functions at the bottom of this file, outside the describe, e.g.:

```javascript

describe('Confirm rounding is consistent across UI, API, and exports', () => {
  // ...
});

function checkAPIDataAgainstDocsTestData = ({student, contentArea, competency, baseline, growth, progress, performanceLevel}) {

}
```

Such functions will get "hoisted" to existing throughout the file even though they're at the bottom. This will make the `describe` block more readable by reducing the testing

**suggestion (non-blocking)**: I'd remove `DocsTestData` as future maintainers won't be aware of the doc we drew test cases out of and it will confusing extra terminology. So this might just be `checkAPIDataAgainstTestCase`

**Review by themightychris — APPROVED** [review]

File: cypress/integration/precise-rounding.js
Comment (themightychris) [reply cypress/integration/precise-rounding.js]: @Camus1859 also note how I quoted the numbers as stings. JavaScript has a lot of quirks when it comes to its `Float` type which is what you get whenever you have an unquoted number with a dot in it. Basically JavaScript has no way to handle fixed-precision numbers natively, and the sorts of weird stuff that happens is along the lines of what we're likely to run into comparing fixed precision numbers of different roundings. Our best bet is to keep our test case data as strings, capture actuals without ever converting them away from strings (in the case of UI and CSV) or turn them into strings immediately and never touch them with math operators (in the case of JSON responses), and do strict equality checks on strings vs strings so JavaScript is never tempted to coerse types

File: cypress/integration/precise-rounding.js
Comment (themightychris) [inline cypress/integration/precise-rounding.js]: I _think_ the spreadsheets should have 0, not blank, for progress values where the value is 0. The only case where blank would make sense is if progress in null because a given competency has 0 demonstration requirements for whatever reason. Otherwise it should also have a numeric value and we should render 0's where a calculation could be made and resulted in 0

Is that right @tgaffey ? would we break any of your processes to start having 0s instead of blanks in progress values?

File: php-classes/Slate/CBL/StudentCompetency.php
Comment (themightychris) [inline php-classes/Slate/CBL/StudentCompetency.php]: @Camus1859 can we delete this line before merging? or even better: delete/edit the commit that added it

Comment (themightychris) [reply php-classes/Slate/CBL/StudentCompetency.php]: This change is still showing under GitHub's Files Changed tab so its still in the PR

File: cypress/integration/precise-rounding.js
Comment (themightychris) [inline cypress/integration/precise-rounding.js:85]: @Camus1859 forget what I said on the call about not using selectors with generated IDs, this code is resolving those dynamically based on stable selectors

**Comment by nbey** [comment]
Looks like all tests are passing, except for blog tests. I've created a PR that should hopefully address those issues here: https://github.com/JarvusInnovations/emergence-skeleton/pull/309 otherwise, I think it looks good to merge.

**Review by nbey — APPROVED** [review]

**Pull request merged**
